src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp

```diff
@@ -0,0 +1,300 @@
+#include <Common/logger_useful.h>
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_USE_CACHE;
+}
+
+CachedObjectStorage::CachedObjectStorage(
+    ObjectStoragePtr object_storage_,
+    FileCachePtr cache_,
+    const FileCacheSettings & cache_settings_,
+    const std::string & cache_config_name_)
+    : object_storage(object_storage_)
+    , cache(cache_)
+    , cache_settings(cache_settings_)
+    , cache_config_name(cache_config_name_)
+    , log(&Poco::Logger::get(getName()))
+{
+    cache->initialize();
+}
+
+FileCache::Key CachedObjectStorage::getCacheKey(const std::string & path) const
+{
+    return cache->hash(path);
+}
+
+String CachedObjectStorage::getCachePath(const std::string & path) const
+{
+    FileCache::Key cache_key = getCacheKey(path);
+    return cache->getPathInLocalCache(cache_key);
+}
+
+std::string CachedObjectStorage::generateBlobNameForPath(const std::string & path)
+{
+    return object_storage->generateBlobNameForPath(path);
+}
+
+ReadSettings CachedObjectStorage::patchSettings(const ReadSettings & read_settings) const
+{
+    ReadSettings modified_settings{read_settings};
+    modified_settings.remote_fs_cache = cache;
+
+    if (FileCache::isReadOnly())
+        modified_settings.read_from_filesystem_cache_if_exists_otherwise_bypass_cache = true;
+
+    return IObjectStorage::patchSettings(modified_settings);
+}
+
+void CachedObjectStorage::startup()
+{
+    object_storage->startup();
+}
+
+bool CachedObjectStorage::exists(const StoredObject & object) const
+{
+    return object_storage->exists(object);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> CachedObjectStorage::readObjects( /// NOLINT
+    const StoredObjects & objects,
+    const ReadSettings & read_settings,
+    std::optional<size_t> read_hint,
+    std::optional<size_t> file_size) const
+{
+    assert(!objects[0].getPathKeyForCache().empty());
+
+    auto modified_read_settings = patchSettings(read_settings);
+    auto implementation_buffer = object_storage->readObjects(objects, modified_read_settings, read_hint, file_size);
+
+    /// If underlying read buffer does caching on its own, do not wrap it in caching buffer.
+    if (implementation_buffer->isIntegratedWithFilesystemCache()
+        && modified_read_settings.enable_filesystem_cache_on_lower_level)
+    {
+        return implementation_buffer;
+    }
+    else
+    {
+        if (!file_size)
+            file_size = implementation_buffer->getFileSize();
+
+        auto implementation_buffer_creator = [objects, modified_read_settings, read_hint, file_size, this]()
+        {
+            return std::make_unique<BoundedReadBuffer>(
+                object_storage->readObjects(objects, modified_read_settings, read_hint, file_size));
+        };
+
+        if (objects.size() != 1)
+            throw Exception(ErrorCodes::CANNOT_USE_CACHE, "Unable to read multiple objects, support not added");
```

Comment:
```suggestion
            throw Exception(ErrorCodes::CANNOT_USE_CACHE, "Read of multiple objects is unsupported");
```
But does it mean that `*Log` tables (which do appends) will not work?

Reply:
Well, I think I can remove this exception and multiple objects will work fine, just cache key will be deduces from the first path only. Anyway this logic is not applied for s3 and for it we have `buffer->isIntegratedWithFIlesystemCache() == 1` and another implementation is used. For non-s3 I should add a test for `*Log`.